Index: src/Pugs/AST/Internals/Instances.hs
===================================================================
--- src/Pugs/AST/Internals/Instances.hs	(Revision 9727)
+++ src/Pugs/AST/Internals/Instances.hs	(Arbeitskopie)
@@ -704,6 +704,7 @@
     showPerl5 (VStr aa) = showP5ArrayObj "VStr" [showPerl5 aa]
     showPerl5 (VList aa) = showP5ArrayObj "VList" [showPerl5 aa]
     showPerl5 (VType aa) = showP5ArrayObj "VType" [showPerl5 aa]
+    showPerl5 (VCode aa) = showP5Class "VUndef"  -- hack
 
 instance JSON Val where
     showJSON (VUndef) = showJSScalar "VUndef"
